library/ssl_tls13_server.c

```diff
@@ -24,16 +24,957 @@
+                                       const unsigned char *end )
+{
+    int ret = 0;
+    size_t cookie_len;
+    unsigned char const *p = buf;
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "parse cookie extension" ) );
+
+    if( ssl->conf->f_cookie_check != NULL )
+    {
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+        cookie_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+        p += 2;
+
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, cookie_len );
+
+        MBEDTLS_SSL_DEBUG_BUF( 3, "Received cookie", p, cookie_len );
+
+        if( ssl->conf->f_cookie_check( ssl->conf->p_cookie,
+                                       p, cookie_len, ssl->cli_id,
+                                       ssl->cli_id_len ) != 0 )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 2, ( "cookie verification failed" ) );
+            ret = MBEDTLS_ERR_SSL_HRR_REQUIRED;
+        }
+        else
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 2, ( "cookie verification passed" ) );
+        }
+    }
+    else
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "cookie verification skipped" ) );
+    }
+
+    return( ret );
+}
+#endif /* MBEDTLS_SSL_COOKIE_C */
+
+/*
+ *
+ * STATE HANDLING: ClientHello
+ *
+ * There are three possible classes of outcomes when parsing the CH:
+ *
+ * 1) The CH was well-formed and matched the server's configuration.
+ *
+ *    In this case, the server progresses to sending its ServerHello.
+ *
+ * 2) The CH was well-formed but didn't match the server's configuration.
+ *
+ *    For example, the client might not have offered a key share which
+ *    the server supports, or the server might require a cookie.
+ *
+ *    In this case, the server sends a HelloRetryRequest.
+ *
+ * 3) The CH was ill-formed
+ *
+ *    In this case, we abort the handshake.
+ *
+ */
+
+/*
+ * Overview
+ */
+
+/* Main entry point from the state machine; orchestrates the otherfunctions. */
+static int ssl_client_hello_process( mbedtls_ssl_context *ssl );
+
+static int ssl_client_hello_parse( mbedtls_ssl_context *ssl,
+                                   const unsigned char *buf,
+                                   const unsigned char *end );
+
+/* Update the handshake state machine */
+static int ssl_client_hello_postprocess( mbedtls_ssl_context *ssl,
+                                         int hrr_required );
+
+/*
+ * Implementation
+ */
+
+#define SSL_CLIENT_HELLO_OK           0
+#define SSL_CLIENT_HELLO_HRR_REQUIRED 1
+
+static int ssl_client_hello_process( mbedtls_ssl_context *ssl )
+{
+
+    int ret = 0;
+    int hrr_required = SSL_CLIENT_HELLO_OK;
+    unsigned char* buf = NULL;
+    size_t buflen = 0;
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse client hello" ) );
+
+    ssl->major_ver = MBEDTLS_SSL_MAJOR_VERSION_3;
+    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_tls13_fetch_handshake_msg(
+                          ssl, MBEDTLS_SSL_HS_CLIENT_HELLO,
+                          &buf, &buflen ) );
+
+    mbedtls_ssl_tls13_add_hs_hdr_to_checksum( ssl,
```

Comment:
The mbedtls_ssl_add_hs_hdr_to_checksum() is one static function now. it can''t be used now, I just replace it with mbedtls_ssl_add_hs_msg_to_checksum(), so need of mbedtls_ssl_add_hs_hdr_to_checksum() here.

Reply:
Should we change the function type? I think it is wrong

Reply:
No, I didn't change the function type. Originally it will call mbedtls_ssl_add_hs_hdr_to_checksum() firstly, and then update_checksum().  mbedtls_ssl_add_hs_msg_to_checksum() will always firstly call mbedtls_ssl_add_hs_hdr_to_checksum() and then update_checksum(), what I change is: replace the first time call update_checksum() with mbedtls_ssl_add_hs_msg_to_checksum(), then it will implement the same function.

Reply:
> Should we change the function type? I think it is wrong

Which function type is changed?